# <span style="color:blue"> MBA em Ciência de Dados</span>
# <span style="color:blue">Técnicas Avançadas para Captura e Tratamento de Dados</span>

## <span style="color:blue"> Web Scraping </span>

## <span style="color:blue">Avaliação </span> 

**Material Produzido por Luis Gustavo Nonato**<br>
**Cemeai - ICMC/USP São Carlos**

---

In [17]:
import requests as rq
from bs4 import BeautifulSoup

Considere o portal de dados abertos do Governo Federal: 

[http://www.dados.gov.br/](http://www.dados.gov.br/)

In [18]:
url = 'http://www.dados.gov.br/dataset'

dados = rq.get(url, params={'q': 'seguranca'})
print('Status do request: ', dados.status_code)

Status do request:  200


In [19]:
dados_seg = BeautifulSoup(dados.text, 'html.parser')

### Exercício 1)
Faça uma requisição no side de dados abertos do Governo Federal utilizando a palavra "seguranca" (sem ç) como fonte de busca. Dentre os repositórios que aparecem na primeira página resultante da busca, quantos caracteres aparecem no nome do repositório com o nome mais longo?

Por exemplo, o nome do repositório
```html
Segurança Pública com Cidadania

Segurança Pública com Cidadania. Empreendimento pertencente a carteira do Anexo III do PPA 2012-2015. Programa 2070. Iniciativas 03DX e 04AD - escala 1:2.500.000. Sistema de...
```
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<span style="background-color:blue">HTML</span>
<span style="background-color:orange">XML</span>

é _"Segurança Pública com Cidadania"_ e possui 31 caracteres.

a) 77<br>
b) 78<br>
c) 79<br>
d) 80

**Dica**: Os repositórios estão marcados pela `tag` 
```html
  <li class="dataset-item">
```

In [20]:
name_rep = {}
for repositorio in dados_seg('li', {'class': 'dataset-item'}):
    titulo = repositorio.find('h3').text
    data = len(titulo.strip())
    print(f'{titulo.strip()} ==> {data}')
    name_rep[titulo.strip()] = data
    print(5*'--')

Segurança Pública com  Cidadania ==> 32
----------
Segurança Pública com Cidadania ==> 31
----------
Indicadores sobre Segurança Pública ==> 35
----------
Regiões Integradas de Segurança Pública ==> 39
----------
Seminário: Fronteiras e Segurança Nacional ==> 42
----------
Seminário: Fronteiras e Segurança Nacional ==> 42
----------
Indicadores sobre Segurança Pública (editado) ==> 45
----------
Equipamentos Públicos de Segurança Alimentar e Nutricional ==> 58
----------
Municípios com Conselho Municipal de Segurança Alimentar ==> 56
----------
Diagnostico da Juventude - Segurança e Justiça ==> 46
----------
Indicadores sobre Força Nacional de Segurança Pública ==> 53
----------
MAPASAN - Mapeamento de Segurança Alimentar e Nutricional ==> 57
----------
Indicadores sobre Aprimoramento das Atividades de Segurança Pública ==> 67
----------
Indicadores sobre Adequação de Unidades de Segurança Pública ==> 60
----------
A Opinião Pública e a Segurança do Estado - 2004 ==> 48
----------
A Op

In [21]:
print(sorted(name_rep.items(), key=lambda kv: kv[1], reverse=True)[0])

('Indicadores sobre PRONASCI - Programa Nacional de Segurança Pública com Cidad...', 80)


## RESPOSTA:<br>
**LETRA D: 80**

### Exercício 2) 
Quantos arquivos estão disponíveis para download nos repositórios listados na primeira página retornada pela requisição do exercício 1)?

a) entre 30 e 35<br>
b) entre 100 e 105<br>
c) entre 70 e 75<br>
d) entre 80 e 90

**Dica**: Você deve extrair a URL correspondente à página de cada repositório, fazer uma requisição a tais páginas e contar o número de arquivos listados em cada uma delas. Os arquivos estão marcados pela `tag` no HTML de cada repositório
```html
  <li class="resource-item">
```

In [22]:
link_rep = {}
url = 'http://www.dados.gov.br'

j = 0
for links in dados_seg('li', {'class': 'dataset-item'}):
    titulo = links.find('h3').text
    titulo = titulo.strip()
    link = links.find('a').get('href')
    print(f'{titulo} ==> {link}')

    page = rq.get(url+link)
    # print('Status do request: ', dados.status_code)
    page_bs = BeautifulSoup(page.text, 'html.parser')
    
    i = 0
    for files in page_bs('li', {'class': 'resource-item'}):
        x = files.find('a').get('href')
        i += 1

    j+=i    
    print(f'{titulo} ==> {i} ==> {j}')
    link_rep[titulo] = i

    print(5*'--')

Segurança Pública com  Cidadania ==> /dataset/mpog_seguranca_publica
Segurança Pública com  Cidadania ==> 5 ==> 5
----------
Segurança Pública com Cidadania ==> /dataset/seguranca-publica-com-cidadania
Segurança Pública com Cidadania ==> 2 ==> 7
----------
Indicadores sobre Segurança Pública ==> /dataset/seguranca-publica
Indicadores sobre Segurança Pública ==> 4 ==> 11
----------
Regiões Integradas de Segurança Pública ==> /dataset/regioes-integradas-de-seguranca-publica
Regiões Integradas de Segurança Pública ==> 3 ==> 14
----------
Seminário: Fronteiras e Segurança Nacional ==> /dataset/seminario-fronteiras-e-seguranca-nacional1
Seminário: Fronteiras e Segurança Nacional ==> 1 ==> 15
----------
Seminário: Fronteiras e Segurança Nacional ==> /dataset/seminario-fronteiras-e-seguranca-nacional
Seminário: Fronteiras e Segurança Nacional ==> 1 ==> 16
----------
Indicadores sobre Segurança Pública (editado) ==> /dataset/seguranca-publica--editado
Indicadores sobre Segurança Pública (edita

In [23]:
print('Foram encontrados {} links na Primeira página'.format(j))

Foram encontrados 100 links na Primeira página


## RESPOSTA:<br>
**LETRA B: ENTRE 100 E 105**

### Exercício 3)
Repita o exercício 2), porém, contando apenas os aquivos de formato CSV que estão disponíveis para download. O total de arquivos CSV é:

a) entre 20 e 25<br>
b) entre 40 e 45<br>
c) entre 30 e 35<br>
d) entre 50 e 55<br>

In [24]:
link_csv = {}
url = 'http://www.dados.gov.br'

j = 0
for links in dados_seg('li', {'class': 'dataset-item'}):
    titulo = links.find('h3').text
    titulo = titulo.strip()
    link = links.find('a').get('href')
    print(f'{titulo} ==> {link}')

    page = rq.get(url+link)
    # print('Status do request: ', dados.status_code)
    page_bs = BeautifulSoup(page.text, 'html.parser')
    
    i = 0
    for files in page_bs('li', {'class': 'resource-item'}):
        y = files.find('span').get('data-format')
        if (y == 'csv') | (y == 'csv+zip'): 
            x = files.find('a').get('href')
            i += 1

    j+=i    
    print(f'{titulo} ==> {i} ==> {j}')
    link_csv[titulo] = i

    print(5*'--')

Segurança Pública com  Cidadania ==> /dataset/mpog_seguranca_publica
Segurança Pública com  Cidadania ==> 1 ==> 1
----------
Segurança Pública com Cidadania ==> /dataset/seguranca-publica-com-cidadania
Segurança Pública com Cidadania ==> 0 ==> 1
----------
Indicadores sobre Segurança Pública ==> /dataset/seguranca-publica
Indicadores sobre Segurança Pública ==> 0 ==> 1
----------
Regiões Integradas de Segurança Pública ==> /dataset/regioes-integradas-de-seguranca-publica
Regiões Integradas de Segurança Pública ==> 0 ==> 1
----------
Seminário: Fronteiras e Segurança Nacional ==> /dataset/seminario-fronteiras-e-seguranca-nacional1
Seminário: Fronteiras e Segurança Nacional ==> 0 ==> 1
----------
Seminário: Fronteiras e Segurança Nacional ==> /dataset/seminario-fronteiras-e-seguranca-nacional
Seminário: Fronteiras e Segurança Nacional ==> 0 ==> 1
----------
Indicadores sobre Segurança Pública (editado) ==> /dataset/seguranca-publica--editado
Indicadores sobre Segurança Pública (editado) 

In [25]:
print('Foram encontrados {} links com arquivos CSV'.format(j))

Foram encontrados 32 links com arquivos CSV


## RESPOSTA:<br>
**LETRA C: ENTRE 30 E 35**

### Exercício 4) 
Um total de 11 páginas são retornadas como resultado da requisição realizada no exercício 1). Percorra os nomes dos repositórios em todas as 11 páginas e calcule quantos repositórios **não** possuem a palavra "segurança" como parte do nome. O número encontrado é:

a) entre 185 e 195<br>
b) entre 170 e 175<br>
c) entre 150 e 155<br>
d) entre 75 e 80<br>


**Obs:** A busca com a palavra "seguranca" é feita sem o ç, porém a busca nos nomes dos repositórios pela palavra "segurança" tem que ser feita com o ç. Não faça distinção entre letras maiúsculas ou minúsculas, ou seja, "Segurança" e "segurança" devem ser consideradas na contagem.

**Dica:** Você consegue navegar pelas páginas enviando o parâmetro
```python
params={'page':pagina,'q': 'seguranca'}
```
para o comando <font color='blue'>get</font>, onde `pagina` é o número da página que se deseja buscar.

In [26]:
import re

url = 'http://www.dados.gov.br/dataset'

token = '(S|s)(eguran)([\w+\W+])'
j = 0
for pagina in range(1, 12):
    dados = rq.get(url, params={'page':pagina,'q': 'seguranca'})
    # print('Status do request: ', dados.status_code)

    dados_pag = BeautifulSoup(dados.text, 'html.parser')
    
    i = 0
    for pagina_name in dados_pag('li', {'class': 'dataset-item'}):
        titulo = pagina_name.find('h3').text
        titulo = titulo.strip()
        if len(re.findall(token, titulo)) == 0:
            # print(re.findall("S(EGURAN)[\w\W]", titulo))
            i =1
        j+=i
        print(f'Página: {pagina} ==> {titulo} ==> {i} ==> {j}')

Página: 1 ==> Segurança Pública com  Cidadania ==> 0 ==> 0
Página: 1 ==> Segurança Pública com Cidadania ==> 0 ==> 0
Página: 1 ==> Indicadores sobre Segurança Pública ==> 0 ==> 0
Página: 1 ==> Regiões Integradas de Segurança Pública ==> 0 ==> 0
Página: 1 ==> Seminário: Fronteiras e Segurança Nacional ==> 0 ==> 0
Página: 1 ==> Seminário: Fronteiras e Segurança Nacional ==> 0 ==> 0
Página: 1 ==> Indicadores sobre Segurança Pública (editado) ==> 0 ==> 0
Página: 1 ==> Equipamentos Públicos de Segurança Alimentar e Nutricional ==> 0 ==> 0
Página: 1 ==> Municípios com Conselho Municipal de Segurança Alimentar ==> 0 ==> 0
Página: 1 ==> Diagnostico da Juventude - Segurança e Justiça ==> 0 ==> 0
Página: 1 ==> Indicadores sobre Força Nacional de Segurança Pública ==> 0 ==> 0
Página: 1 ==> MAPASAN - Mapeamento de Segurança Alimentar e Nutricional ==> 0 ==> 0
Página: 1 ==> Indicadores sobre Aprimoramento das Atividades de Segurança Pública ==> 0 ==> 0
Página: 1 ==> Indicadores sobre Adequação de U

In [27]:
print('Foram encontrados {} nomes sem a palavra "SEGURANÇA"'.format(j))

Foram encontrados 190 nomes sem a palavra "SEGURANÇA"


## RESPOSTA:<br>
**LETRA A: ENTRE 185 E 190**

### Exercício 5) 
Os repositórios retornados pela requisição do exercício 1) podem trazer a informação sobre o "Autor" do repositório, ou seja, quem foi o criador do repositório. Percorra as 11 páginas retornadas da requisição do exercício 1) e, para todos os repositórios listados em cada página, verifique qual é o "Autor" que aparece mais frequentemente como criador dos repositórios. 

O "Autor" mais frequente é: 

a) 'GSI' com aproximadamente 68 repositórios<br>
b) 'SNJ' com aproximadamente 27 repositórios<br>
c) 'Ministério da Justiça' com aproximadamente 16 repositórios<br>
d) 'SNJ' com aproximadamente 76 repositórios<br>

In [31]:
link_autor = {}
url = 'http://www.dados.gov.br'

for pagina in range(1, 12):
    dados = rq.get(url, params={'page':pagina,'q': 'seguranca'})
    # print('Status do request: ', dados.status_code)

    dados_pag = BeautifulSoup(dados.text, 'html.parser')
    
    print(10*'**')
    print(f'{pagina}')
    print(10*'**')
    
    for links in dados_pag('li', {'class': 'dataset-item'}):
        titulo = links.find('h3').text
        titulo = titulo.strip()
        link = links.find('a').get('href')
        # print(f'{titulo} ==> {link}')

        page = rq.get(url+link)
        print('Status do request: ', dados.status_code)
        page_bs = BeautifulSoup(page.text, 'html.parser')
        
        for files in page_bs('td', {'property': 'dc:creator'}):
            try:
                autor = files.find('a').text
                autor = autor.strip()
                print('try')
                if autor in link_autor:
                    link_autor[autor] += 1
                else:
                    link_autor[autor] = 1

            except:
                autor = files.text
                autor = autor.strip()
                print('except')
                if autor in link_autor:
                    link_autor[autor] += 1
                else:
                    link_autor[autor] = 1

            print(f'{titulo} ==> {autor}')

        print(5*'--')

********************
1
********************
Status do request:  200
except
Microdados do Censo Escolar ==> Instituto Nacional de Estudos e Pesquisas Educacionais Anísio Teixeira - INEP
----------
Status do request:  200
----------
Status do request:  200
try
Orçamento Federal - Orçamento Fiscal e da Seguridade Social ==> Secretaria de Orçamento Federal - SOF/MP
----------
********************
2
********************
Status do request:  200
try
A - Série histórica de preços de combustíveis - revenda ==> Agência Nacional do Petróleo, Gás Natural e Biocombustíveis - ANP
----------
Status do request:  200
try
Sistema Nacional de Informações Florestais - SNIF ==> Ana Laura Cerqueira Trindade
----------
Status do request:  200
try
Serviços Públicos Federais para o Cidadão ==> Ministério do Planejamento
----------
********************
3
********************
Status do request:  200
except
Microdados do Censo Escolar ==> Instituto Nacional de Estudos e Pesquisas Educacionais Anísio Teixeira - IN

KeyboardInterrupt: 

In [29]:
print(sorted(link_autor.items(), key= lambda kv: kv[1], reverse=True))

[('Instituto Nacional de Estudos e Pesquisas Educacionais Anísio Teixeira - INEP', 7), ('Tribunal Superior Eleitoral - TSE', 7), ('Centro de Investigação e Prevenção de Acidentes Aeronáuticos', 7), ('Sala de Apoio à Gestão Estratégica - SAGE', 4), ('Secretaria de Orçamento Federal - SOF/MP', 4), ('Diretoria de Geociências (DGC) / Coordenação de Cartografia  (CCAR)', 4)]


In [30]:
link_autor

{'Instituto Nacional de Estudos e Pesquisas Educacionais Anísio Teixeira - INEP': 7,
 'Tribunal Superior Eleitoral - TSE': 7,
 'Centro de Investigação e Prevenção de Acidentes Aeronáuticos': 7,
 'Sala de Apoio à Gestão Estratégica - SAGE': 4,
 'Secretaria de Orçamento Federal - SOF/MP': 4,
 'Diretoria de Geociências (DGC) / Coordenação de Cartografia  (CCAR)': 4}